ARC: Add support for ARC HS family on nSIM

abrodkin opened the pull request:
Now with generalized nSIM platfrom it's easy to add support of another ARC core family support in Zephyr which we do here.

Just use `BOARD=nsim_hs` instead of `nsim_em`.
Note existing Zephyr SDK is perfectly capable of building binaries for ARC HS cores so we should be all set.

Commit:
ARC: Fix selection of custom atomic ops

Up until now only ARC EM family has been supported in Zephyr
which don't support atomic operations other than
compare-and-excange, so custom atomic ops with load-locked(LLOCK)/
store-conditional(SCOND) were never used that's how we never
realised CONFIG_ATOMIC_OPERATIONS_CUSTOM points to the wrong file:
"atomic.c" while real implementation is in "atomic.S".

Fix that now.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

Commit:
ARC: Add support for ARC HS family of CPU cores

The ARC HS is a family of high performance CPUs from Synopsys
capable of running wide range of applications from heavy DPS
calculation to full-scale OS.

Still as with other ARC cores ARC HS might be tailored to
a particular application.

As opposed to EM cores ARC HS cores always have support of unaligned
data access and by default GCC generates such a data layout with
so we have to always enable unaligned data access in runtime otherwise
on attempt to access such data we'd see "Unaligned memory exception".

Note we had to explicitly mention CONFIG_CPU_ARCEM=y in
all current defconfigs as CPU_ARC{EM|HS} are now parts of a
choice so we cannot simply select ether option in board's Kconfig.

And while at it change "-mmpy-option" of ARC EM to "wlh1"
which is the same as previously used "6" but matches
Programmer's Reference Manual (PRM) and is more human-friendly.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin committed:
board/nsim: Add support of ARC HS cores in nSIM

ARC nSIM simulates pretty much any modern ARC core,
moreover it emulates a lot of different core features so
it is possible to play with them even wo real hardware.

Thus we add yet another ARC core family to be used on simulated
nSIM board.

For now it's just a basic configuration with ARC UART for
smoke-testing of Zephyr on ARC HS CPUs.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin commented:
Pushed v2 with removal of unused `CONFIG_CPU_xxx`.

abrodkin commented:
Pushed v3 with updated docs for nsIM board.

dbkinder commented on boards/arc/nsim/doc/index.rst:
```diff
@@ -7,20 +7,22 @@ Overview
 * a virtual output only console (uart-nsim)
 
-There are two sub configurations in board:
+There are three sub configurations in board:
```
```suggestion
There are three supported board sub-configurations:
```

dbkinder reviewed (changes requested):
one little grammar fix, otherwise docs LGTM

dbkinder reviewed (approved):
doc changes LGTM, thanks!

ruuddw commented on boards/arc/nsim/doc/index.rst:
```diff
@@ -7,20 +7,22 @@ Overview
-* nsim_sem which includes secure em features and ARC MPUv3
+* ``nsim_em`` which includes normal ARC EM features and ARC MPUv2
+* ``nsim_sem`` which includes secure ARC EM features and ARC MPUv3
+* ``nsim_hs`` which includes normal ARC HS features
```
what does 'normal' mean in HS context? Guess MMU and MPU are not (yet) supported?
Maybe call it 'base' instead of normal.

abrodkin replied:
Correct! It was just a copy-paste of EM's description.

So what if I say:
```
nsim_hs which includes base ARC HS features, i.e. without PMU & MMU
```
?

ruuddw reviewed (changes requested):
Few minor remarks inline, but please add a 'nsim_hs.yaml' in the boards/arc/nsim directory so sanitycheck can be used for testing (sanitycheck -p nsim_hs).
Btw, I added and tested this, result looks good:
132 of 134 tests passed with 0 warnings

abrodkin commented on arch/arc/Kconfig:
```diff
@@ -12,17 +12,25 @@ menu "ARC Options"
+config	CPU_ARCHS
+	bool "ARC HS cores"
+	select CPU_ARCV2
+	select ATOMIC_OPERATIONS_BUILTIN
```
My idea was if such not very straight-forward code was never really used up until now (see we mentioned even wrong file name, https://github.com/zephyrproject-rtos/zephyr/pull/16662/commits/2f7990c8b5a012cc0484ea8ce3ab40dbc721b148) it might be a better idea to use atomic ops provided by the GCC which are known to work (even though I don't know any directed stress test for GCC's atomics) as they are used pretty often in complex software packages like Qt (and many others). At least for starters. And then once we make sure those hand-written atomics do work OK we may switch ARC HS cores to use custom version by default. That way we'll be moving with smaller steps rather than adding many major changes in 1 merge.

What if I create a new issue for switch ARC HS to custom atomics? But I guess it makes sense to do it after HS addition is merged in the master branch.

andrewboie replied:
Honestly, it's better to use the compiler builtins if they are available.

abrodkin replied:
@andrewboie well it's up to the user's philosophy. I mean in the Linux kernel and U-Boot bootloader we prefer to have as few external dependencies as possible thus trying to have everything we need from libgcc, string functions etc right in the kernel's source tree.

Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.

But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that as it's much easier and allows us to reduce amount of hand-crafted assembly (which might very well have subtle issues since we're just human beings).

andrewboie replied:
> Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.

Not sure what you mean by this.
The semantics of Zephyr's atomic APIs should be the same across arches. Otherwise this would break arch-agnostic code that uses them (futexes, etc)

> But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that 

For all our other arches, we prefer builtins if they exist. Arches that use the C implementations or custom assembly, AFAIK, all do this because the builtins are not implemented or not working for some reason.

abrodkin replied:
> > Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.
> 
> Not sure what you mean by this.
> The semantics of Zephyr's atomic APIs should be the same across arches. Otherwise this would break arch-agnostic code that uses them (futexes, etc)

Indeed. I meant compiler built-in atomics might have semantics which differ from what one may expect in his particular project. Moreover I guess different compilers (not sure if that's applicable to the Zephyr today as probably only GCC is supposed to be used, but what if at some point other compilers are adopted like Clang/LLVM, proprietary ones like IAR or Keil etc) and even compiler versions might have a bit different semantics of atomics.

But probably I'm exaggerating here.
 
> > But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that
> 
> For all our other arches, we prefer builtins if they exist. Arches that use the C implementations or custom assembly, AFAIK, all do this because the builtins are not implemented or not working for some reason.

That's a good move and as you may see I'm all for using built-ins as more robust and simple solution. Then do you suggest to get rid of custom atomic ops for ARC at all?

andrewboie replied:
>  Then do you suggest to get rid of custom atomic ops for ARC at all?

If we can, yes.
My memory is hazy, but I think the reason why those ASM implementations exist is because GCC didn't have builtins for ARC when the ARC port was added to Zephyr, but I'd have to investigate more to be sure.

andrewboie replied:
> Indeed. I meant compiler built-in atomics might have semantics which differ from what one may expect in his particular project. Moreover I guess different compilers (not sure if that's applicable to the Zephyr today as probably only GCC is supposed to be used, but what if at some point other compilers are adopted like Clang/LLVM, proprietary ones like IAR or Keil etc) and even compiler versions might have a bit different semantics of atomics.

Ah I see. Yeah if a compiler has different atomic semantics then there would need to be a separate set of wrappers to the Zephyr atomic APIs. I expect LLVM would work the same as GCC, but who knows with other compilers.

abrodkin replied:
> 
> 
> > Then do you suggest to get rid of custom atomic ops for ARC at all?
> 
> If we can, yes.


Ok will do it shortly as otherwise we do need to have a good and thorough study of custom implementation which I'd like to escape :)

> My memory is hazy, but I think the reason why those ASM implementations exist is because GCC didn't have builtins for ARC when the ARC port was added to Zephyr, but I'd have to investigate more to be sure.

I took a look already and at least from this git repo I see ARC custom atomics were added in the very first commit 8ddf82cf70dc6f951ab477f325dee0efde3ec589. 

What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to https://stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.

andrewboie replied:
> What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.

You get linker errors if the atomics aren't actually implemented for that arch.
I ran into this when I wrote the Nios II port.

abrodkin replied:
> 
> 
> > What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.
> 
> You get linker errors if the atomics aren't actually implemented for that arch.
> I ran into this when I wrote the Nios II port.

Indeed! Atomic ops for ARC were only added in the end of 2015 (https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/f5599934783a78e3e89c69a88c6cc2fe741c7afa) and made it to arc-2016.03 release which was published on December 9-th 2016 (https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/releases/tag/arc-2016.09-release). which was much later than the first commit in Zephyr git repo :)

Mystery solved!

abrodkin commented:
The following changes were done in the re-spin:
1. Added sanity check scenario for `nsim_hs` (run `./scripts/sanitycheck -p nsim_hs`)
2. Updated description of nSIM platform with ARC HS
3. Added ARC HS specific `-mmpy-option` to accommodate 54b229fc422d9dc8f57b765d891fb3d0bdbf4a9f

abrodkin committed:
arc: Add support for unaligned access

ARCv2 cores may access data not aligned by the data size boundary.
I.e. read entire 32-bit word from address 0x1.

This feature is configurable for ARC EM cores excluding those with
secure shield 2+2 mode. When it's available in hardware it's required
to enable that feature in run-time as well setting status32.AD bit.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin committed:
arc: Preserve STATUS32 flags while resetting AE flag

KFLAG instruction might affect multiple flags in STATUS32 register
and so when we need just AE-bit to be reset we need first read current
state of STATUS32, then change our bit and set STATUS32 again.

Otherwise critical flags including stack checking, unaligned access etc
will be dropped for good.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin committed:
boards/nsim: Enable unaligned data acess for nSIM with simple ARC EM

This will give us a possibility to check unaligned read/write support
in simulation.

Note nSIM with S(ecure)EM (with secure option) doesn't support that
mode in HW.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin commented:
In the last re-spin the following changes were done:
 * Added support for unaligned access strictly required for ARC HS cores and optional for ARC EM w/o security option (enabled by default for nSIM with ARC EM, AKA `nsim_em`). This was required to pass full sanity check test set for `nsim_hs` w/o any failures.
 * To properly support enabled unaligned access fixed STATUS32 tinkering code

vonhust commented:
This PR only can be merged after PR #17420 is merged

abrodkin commented:
@ruuddw could you please check if your comments were all addressed and mark corresponding conversation as solved so that this PR becomes ready for merge?

vonhust commented:
@ruuddw, this PR  is next one to merge.  Then we can go to arc smp->HSDK

abrodkin commented:
@andrewboie may we merge this one as a prerequisite for SMP for ARC (https://github.com/zephyrproject-rtos/zephyr/pull/17747)?
